app/models/ability.rb

```diff
@@ -204,18 +206,22 @@ def item_abilities
   end
 
   def message_abilities
-    # Message (sender and admins, not user if private is true)
     if can_manage_messages?
-      can [:index, :show, :create, :update, :destroy], Message
-    elsif can_create_and_read_messages?
-      can [:index, :show, :create], Message
+      # Supervisors and Reviewers can see all messages on all offers (and items belonging to an offer)
+      # Order fulfillment and Order Administrators are allowed to see all messages on all orders.
+      can %i[index show create], Message
     else
-      can [:index, :show, :create], Message, Message.donor_messages(@user_id) do |message|
+      # A donor / charity user is allowed to read any non-private message
+      # that is on an offer / order created by them
+      can [:index, :show], Message, @user.messages do |message|
```

Comment:
Not quite... we need to cover the following cases:

1. Donor can `index, show` messages they created: `message.sender_id == donor.id`
2. Donor can `index, show` messages sent to them: `message.subscriptions.map(&:user_id).include(donor.id)`

The way we did this before was with the `donor_messages` method which scoped to non-private messages on offers created by the donor. We need to expand this to include orders.

The problem with the code above is that `@user.messages` is only messages created by the user and doesn't include messages sent to them.

Taking a step back, I think we can simplify this. We need a list of all messages a user is allowed to see. We already control this with the `subscriptions` table. So can't we just do this?

```
class Ability

  def message_abilities
    can [:index, :show, :create], Message, messageable_type: 'Offer'  if can_manage_offer_messages?
    can [:index, :show, :create], Message, messageable_type: 'Item'  if can_manage_offer_messages?
    can [:index, :show, :create], Message, messageable_type: 'Order'  if can_manage_order_messages?
    can [:index, :show, :mark_read, :mark_all_read], Message, id: @user.subscriptions.pluck(:message_id)

    # Normal users can create non private messages on objects they own
    can :create, Message do |message|
      message.related_object&.created_by_id == @user_id && !message.is_private && @user_id != nil
    end

end
```

Note, I've also split `can_manage_messages` into `can_manage_offer_messages` and `can_manage_order_messages`. Seems a sensible way to distinguish between staff working on Offers and staff working on Orders.

Feels way too simple... am I missing a case? The specs should verify that.